finality-aleph/src/finalization.rs

```diff
@@ -107,6 +119,181 @@ where
     None
 }
 
+pub(crate) struct FinalizableChain<BE, B, C, St>
+where
+    B: Block,
+    BE: Backend<B>,
+    C: crate::ClientForAleph<B, BE>,
+    St: Stream<Item = B::Hash> + Unpin,
+{
+    hashes: St,
+    last_finalized_hash: B::Hash,
+    client: Arc<C>,
+    max_h: NumberFor<B>,
+    finished: bool,
+    phantom: PhantomData<BE>,
+}
+
+impl<
+        B: Block,
+        BE: Backend<B>,
+        C: crate::ClientForAleph<B, BE>,
+        St: Stream<Item = B::Hash> + Unpin,
+    > FinalizableChain<BE, B, C, St>
+{
+    pub(crate) fn new(client: Arc<C>, hashes: St, max_h: NumberFor<B>) -> Self {
+        let last_finalized_hash = client.info().finalized_hash;
+        let finished = client.info().finalized_number >= max_h;
+        FinalizableChain {
+            hashes,
+            last_finalized_hash,
+            client,
+            max_h,
+            finished,
+            phantom: PhantomData,
+        }
+    }
+    pub(crate) async fn next_hash(&mut self) -> Option<B::Hash> {
+        if self.finished {
+            return None;
+        }
+        loop {
+            let hash = self.hashes.next().await?;
+            let reduced = match reduce_block_up_to(self.client.clone(), hash, self.max_h) {
+                Some(reduced) => reduced,
+                None => continue,
+            };
+            let lca = sp_blockchain::lowest_common_ancestor(
+                self.client.as_ref(),
+                reduced,
+                self.last_finalized_hash,
+            )
+            .map_err(|e| error!(target: "afa", "unable to find lca: {:?}", e))
+            .ok()?;
+            if lca.hash != reduced {
```

Comment:
this should be `if lca.hash != self.last_finalized_hash`, but after fixing that, the blocks still do not finalize, I am still investingating why